File: triangle.py
class Row:
    def __init__(self, rowno):
        # Preallcate row cells
        self.data = [0 for _ in range(rowno + 1)]
        nspaces = rowno
        self.str_len = nspaces

class Triangle:
    def __init__(self, height):
        self.height = height
        self.rows = [Row(rowno) for rowno in range(height)]
        for rowno in range(height):
            for colno in range(rowno + 1):
                row_cur = self.rows[rowno]
                row_cur.data[colno] =\
                    1 if rowno == 0 or colno == 0 or colno == rowno\
                      else sum(self.rows[rowno - 1].data[colno - 1:colno + 1])
                row_cur.str_len += len(str(row_cur.data[colno]))

    def __str__(self):
        t_str = ''
        base_str_len = self.rows[self.height - 1].str_len
        for rowno in range(self.height):
            row_cur = self.rows[rowno]
            pad_len = (base_str_len - row_cur.str_len) // 2
            t_str += ' ' * pad_len
            for colno in range(rowno + 1):
                t_str += str(row_cur.data[colno])
                t_str += ' ' if colno < rowno else '\n'

        return t_str

if __name__ == '__main__':
    height = input('Enter height: ')
    t = Triangle(int(height))
    print(str(t))

__all__ = [\
  'Triangle',\
]

File: triangle.cpp
#include <iostream>
#include <cstddef>
#include <vector>
#include <sstream>

struct Row
{
    std::vector<std::size_t> data;
    std::size_t str_len;

    std::size_t &operator[](std::size_t);
};

std::size_t &
Row::operator[](std::size_t colno)
{
    return data[colno];
}

struct Triangle
{
    std::size_t height;
    std::vector<Row> rows;

    Triangle(std::size_t);

    std::string str(void);
};

Triangle::Triangle(std::size_t height) :
    height {height}, rows {height}
{
    for (std::size_t rowno {0}; rowno < height; ++rowno)
    {
        Row &row = rows[rowno];
        row.data.reserve(rowno + 1);
        for (std::size_t colno {0}; colno <= rowno; ++colno)
        {
            row[colno] =
                (rowno == 0 || colno == 0 || colno == rowno) ?
                1 :
                rows[rowno - 1][colno - 1] + rows[rowno - 1][colno];
            row.str_len += std::to_string(row[colno]).size();
        }
        row.str_len += rowno;
    }
}

std::string
Triangle::str(void)
{
    std::ostringstream oss {};
    Row &base_row = rows[height - 1];
    for (std::size_t rowno {0}; rowno < height; ++rowno)
    {
        Row &row = rows[rowno];

        // Pad on the left for centering
        std::size_t pad_len 
        {
            (base_row.str_len - row.str_len) / 2
        };
        for (std::size_t i = 0; i < pad_len; ++i)
        {
            oss << " ";
        }

        // Add elements
        for (std::size_t colno {0}; colno <= rowno; ++colno)
        {
            oss << row[colno];
            oss << (colno == rowno ? "\n" : " ");
        }
    }
    return oss.str();
}

int 
main()
{
    Triangle t {15};
    std::cout << t.str();
    return 0;
}

